### PR TITLE
Handle case of double "-" chars in input component

### LINKED ISSUE #161: Handling of negative values in input component
Fix issue for token amount input.  The negative values for input should be handle correctly. The user should not be able to enter negative values or double "-" chars. More information here https://github.com/tahowallet/dapp/pull/156.

image

### PR DESCRIPTION
**Pull request by xpaczka**
Resolves #161 

## What has been done:
Added `regex` to check for number values, "-" and empty value. Also added `acceptLetters` prop to not run `onChange` function when we do not want to accept letters in the input and value must match the regex

## Testing
- [x] Enter "0" - Stake button should be disabled
- [x] Enter positive value smaller than balance - Stake button should be enabled
- [x] Enter positive value greater than balance - Stake button should be disabled and error "Not enough funds" should be displayed
- [x] Enter double "-" char - Input should display only one "-" char
- [x] Enter any non-number value - Input should not update
- [x] Enter negative value - Stake button should be disabled and error "Invalid value" should be displayed
- [x] Test staking / unstaking - check if correct values are provided

### CODE REVIEW

File: src/shared/components/TokenAmountInput.tsx [hunk 1]
```diff
@@ -23,13 +23,15 @@ function handleValidate(
...
-      return { error: AmountErrors.NOT_ENOUGH_FUNDS }
-    }
+  if (
+    (parsed === undefined && value) ||
+    (parsed !== undefined && parsed < 0n)
+  ) {
+    return { error: AmountErrors.INVALID_VALUE }
+  }
+
+  if (parsed !== undefined && parsed > balance) {
```
Comment (ioay) [inline src/shared/components/TokenAmountInput.tsx:33]: Let's write it down more clearly, group by conditions not by results(more human-readable):

```
if(!parsed){
  if(value} ...
}
if(parsed){
  if(parsed < 0n) ...
  if(parsed > balance) ...
}
```

**Review by ioay — CHANGES_REQUESTED** [review]

**Comment by ioay** [comment]
When you want to unstake and you will enter the max value manually button is disabled, in the case when you click the max button in the input field - the button `unstake` becomes active.
Screenshot 2023-12-05 at 20 45 23
Screenshot 2023-12-05 at 20 45 41

**Comment by xpaczka** [comment]
> When you want to unstake and you will enter the max value manually button is disabled, in the case when you click the max button in the input field - the button `unstake` becomes active. Screenshot 2023-12-05 at 20 45 23 Screenshot 2023-12-05 at 20 45 41

It is an issue that was there before, I wasn’t able to resolve it, it happens in some cases

File: src/shared/components/TokenAmountInput.tsx [hunk 2]
```diff
@@ -107,12 +109,12 @@ export default function TokenAmountInput({
         <div className="label">{`${label} ${parsedBalance} ${symbol}`}</div>
       )}
       <SharedInput
-        type="number"
         label={inputLabel}
         value={textAmount}
         disabled={disabled}
         onChange={setTextAmount}
         validate={validate}
+        acceptLetters={false}
```
[the remark this answers is not in the document]
Comment (xpaczka) [reply src/shared/components/TokenAmountInput.tsx:117]: The "number" input does not resolve the issue here, it still accepts double "-" char

[the remark this answers is not in the document]
Comment (xpaczka) [reply src/shared/components/TokenAmountInput.tsx:119]: https://stackoverflow.blog/2022/12/26/why-the-number-input-is-the-worst-input/

Cool article, why using "number" input is the worst :)

File: src/shared/components/Input.tsx [hunk 3]
```diff
@@ -41,7 +45,19 @@ export default function SharedInput({
     handleError(value)
   }, [handleError, value])
 
-  const isTypeNumber = type === "number"
+  const handleInputChange = useCallback(
+    (inputValue: string) => {
```
Comment (ioay) [inline src/shared/components/Input.tsx:49]:
```suggestion
    (e: Event) => {
```

**Review by ioay — DISMISSED** [review]
lgtm, just one minor comment

File: src/shared/components/Input.tsx [hunk 4]
```diff
@@ -11,18 +30,9 @@ export default function SharedInput({
...
-  value: string
-  placeholder?: string
-  onChange?: (value: string) => void
-  validate?: (value: string) => { value: unknown } | { error: string }
-  rightComponent?: React.ReactNode
-  style?: React.CSSProperties & Record<string, string>
-}) {
+}: SharedInputProps) {
   const [error, setError] = useState("")
+  const isTypeNumber = useMemo(() => type === "number", [type])
```
Comment (jagodarybacka) [inline src/shared/components/Input.tsx:35]: is useMemo doing anything? both "type" is a string and "isTypeNumber" is a boolean - they shouldn't cause rerenders or trigger changes without actually changing values, right?

**Review by ioay — APPROVED** [review]

File: src/shared/components/Input.tsx [hunk 5]
```diff
@@ -41,7 +51,20 @@ export default function SharedInput({
     handleError(value)
   }, [handleError, value])
 
-  const isTypeNumber = type === "number"
+  const handleInputChange = useCallback(
+    (event: ChangeEvent<HTMLInputElement>) => {
+      const inputValue = event.target.value
+      if (inputValue === "") onChange?.("")
```
Comment (jagodarybacka) [inline src/shared/components/Input.tsx:57]: shouldn't we "return" in this line?

File: src/shared/components/TokenAmountInput.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (Shadowfiend) [reply src/shared/components/TokenAmountInput.tsx]: That `inputmode` trick is nice :eyes:

**Pull request merged**